src/util/GraphStatus.h

```diff
@@ -0,0 +1,546 @@
+    }
+
+    static GraphStatus setInvalidTaskParam() {
+        return GraphStatus(ErrorCode::E_INVALID_TASK_PARA,
+                           getErrorMsg(ErrorCode::E_INVALID_TASK_PARA));
+    }
+
+    static GraphStatus setUserCancel() {
+        return GraphStatus(ErrorCode::E_USER_CANCEL,
+                           getErrorMsg(ErrorCode::E_USER_CANCEL));
+    }
+
+    static GraphStatus setSpaceNotFound(const std::string &name) {
+        return GraphStatus(ErrorCode::E_SPACE_NOT_FOUND,
+                folly::stringPrintf(
+                        getErrorMsg(ErrorCode::E_SPACE_NOT_FOUND).c_str(), name.c_str()));
+    }
+
+    static GraphStatus setTagNotFound(const std::string &name) {
+        return GraphStatus(ErrorCode::E_TAG_NOT_FOUND,
+                folly::stringPrintf(
+                        getErrorMsg(ErrorCode::E_TAG_NOT_FOUND).c_str(), name.c_str()));
+    }
+
+    static GraphStatus setEdgeNotFound(const std::string &name) {
+        return GraphStatus(ErrorCode::E_EDGE_NOT_FOUND,
+                folly::stringPrintf(
+                        getErrorMsg(ErrorCode::E_EDGE_NOT_FOUND).c_str(), name.c_str()));
+    }
+
+    static GraphStatus setIndexNotFound(const std::string &name) {
+        return GraphStatus(ErrorCode::E_INDEX_NOT_FOUND,
+                folly::stringPrintf(
+                        getErrorMsg(ErrorCode::E_INDEX_NOT_FOUND).c_str(), name.c_str()));
+    }
+
+    static GraphStatus setTagPropNotFound() {
+        return GraphStatus(ErrorCode::E_TAG_PROP_NOT_FOUND,
+                           getErrorMsg(ErrorCode::E_TAG_PROP_NOT_FOUND));
+    }
+
+    static GraphStatus setEdgePropNotFound() {
+        return GraphStatus(ErrorCode::E_EDGE_PROP_NOT_FOUND,
+                           getErrorMsg(ErrorCode::E_EDGE_PROP_NOT_FOUND));
+    }
+
+    static GraphStatus setInvalidVid() {
+        return GraphStatus(ErrorCode::E_INVALID_VID,
+                           getErrorMsg(ErrorCode::E_INVALID_VID));
+    }
+
+    static GraphStatus setInternalError(const std::string &reason) {
+        return GraphStatus(ErrorCode::E_INTERNAL_ERROR,
+                folly::stringPrintf(
+                        getErrorMsg(ErrorCode::E_INTERNAL_ERROR).c_str(), reason.c_str()));
+    }
+
+public:
+    ErrorCode getErrorCode() const {
+        return errorCode_;
+    }
+
+    std::string toString() const {
+        return errorMsg_;
+    }
+
+    friend std::ostream& operator<<(std::ostream &os, const GraphStatus &status);
+
+    bool ok() const {
+        return errorCode_ == ErrorCode::SUCCEEDED;
+    }
+
+public:
+    static cpp2::Language         language;
+    static cpp2::Encoding         encoding;
+
+private:
+    static cpp2::Language toLanguage(const std::string &l);
+
+    static cpp2::Encoding toEncoding(const std::string &e);
+
+    static std::string encode(const std::string &errorMsg);
+
+private:
+    ErrorCode        errorCode_{ErrorCode::SUCCEEDED};
```

Comment:
Why not use Status directly?

Reply:
`Status` is an internal error, driven by error message, and cannot be used to enforce the error code returned. if enforce it use error code, internal use is not convenient.

Reply:
I don't think add a new one it better.

Reply:
You can cast internal error when return to user and still is a Status.

Reply:
Cann’t do that. You end up with an internal error, and status shouldn't depend on Thrift.

Reply:
Why shouldn't depend on thrift error code?

Reply:
I also think it's convenient to use single status definition.